From daf6a6d8ad5ebad42dc0040002bf478d135b2051 Mon Sep 17 00:00:00 2001
From: Kim Zhang <xiazhang@codeaurora.org>
Date: Thu, 28 Apr 2016 17:08:00 +0800
Subject: [PATCH 2/3] mm-video-v4l2: vidc: Restore pixel format on capture
 plane

Capture plane's pixel format may be changed during different DPB-OPB modes
switching in real-time, which will lead to a wrong buffer size returned
from driver/fw.
For example, dynamic changing resolution clip with VPP feature enabled.

Change-Id: I12dbed62e87d3934f6630cb1ced9a09ef2a33159
---
 mm-video-v4l2/vidc/vdec/src/omx_vdec_v4l2.cpp | 17 +++++++++++++++++
 1 file changed, 17 insertions(+)

diff --git a/mm-video-v4l2/vidc/vdec/src/omx_vdec_v4l2.cpp b/mm-video-v4l2/vidc/vdec/src/omx_vdec_v4l2.cpp
index c56f097..53f5f71 100644
--- a/mm-video-v4l2/vidc/vdec/src/omx_vdec_v4l2.cpp
+++ b/mm-video-v4l2/vidc/vdec/src/omx_vdec_v4l2.cpp
@@ -981,6 +981,8 @@ OMX_ERRORTYPE omx_vdec::set_dpb(bool is_split_mode, int dpb_color_format)
 OMX_ERRORTYPE omx_vdec::decide_dpb_buffer_mode()
 {
     OMX_ERRORTYPE eRet = OMX_ErrorNone;
+    struct v4l2_format fmt;
+    int rc = 0;
 
     bool cpu_access = capture_capability != V4L2_PIX_FMT_NV12_UBWC;
     bool is_res_above_1080p = false;
@@ -1019,6 +1021,21 @@ OMX_ERRORTYPE omx_vdec::decide_dpb_buffer_mode()
     if (eRet) {
         DEBUG_PRINT_HIGH("Failed to set DPB buffer mode: %d", eRet);
     }
+
+    //Restore the capture format again here, because
+    //in switching between different DPB-OPB modes, the pixelformat
+    //on capture port may be changed.
+    memset(&fmt, 0x0, sizeof(struct v4l2_format));
+    fmt.type = V4L2_BUF_TYPE_VIDEO_CAPTURE_MPLANE;
+    fmt.fmt.pix_mp.height = drv_ctx.video_resolution.frame_height;
+    fmt.fmt.pix_mp.width = drv_ctx.video_resolution.frame_width;
+    fmt.fmt.pix_mp.pixelformat = capture_capability;
+    rc = ioctl(drv_ctx.video_driver_fd, VIDIOC_S_FMT, &fmt);
+    if (rc) {
+        DEBUG_PRINT_ERROR("%s: Failed set format on capture mplane", __func__);
+        return OMX_ErrorUnsupportedSetting;
+    }
+
     return eRet;
 }
 
-- 
1.9.1

